### PR TITLE
CI: enable arbitrary try builds, take two

### PR DESCRIPTION
**Pull request by Kobzol**
Fixed version of https://github.com/rust-lang/rust/pull/124631, which hopefully won't completely break our CI this time :man_facepalming: Sorry once again. Only the last commit is new.

r? @pietroalbini

### CODE REVIEW

File: src/ci/github-actions/calculate-job-matrix.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def calculate_jobs(run_type: WorkflowRunType, job_data: Dict[str, Any]) -> List[
                 raise Exception(f"Custom job(s) `{unknown_jobs}` not found in auto jobs")
 
         return add_base_env(name_jobs(jobs, "try"), job_data["envs"]["try"])
-    elif run_type is AutoRunType:
+    elif isinstance(run_type, AutoRunType):
```
Comment (Kobzol) [inline src/ci/github-actions/calculate-job-matrix.py:139]: This is the bug that caused the PR to run no jobs.

Comment (pietroalbini) [reply src/ci/github-actions/calculate-job-matrix.py:139]: Whoops.

**Pull request merged**